### ISSUE TITLE
II The problem of exploding or vanishing gradients

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#10754]
II The problem of exploding or vanishing gradients<br>
<br>
https://ift.tt/4ltUyBj<br>
<br>
Jingru Guo<br>
<br>

<div><div><div><div><section><h1>I   The importance of effective initialization</h1>
<p>To build a machine learning algorithm, usually you’d define an architecture (e.g. Logistic regression, Support Vector Machine, Neural Network) and train it to learn parameters. Here is a common training process for neural networks:</p>
<ol>
<li>Initialize the parameters</li>
<li>Choose an <span>optimization algorithm</span>
</li>
<li>Repeat these steps:
<ol>
<li>Forward propagate an input</li>
<li>Compute the cost function</li>
<li>Compute the gradients of the cost with respect to parameters using backpropagation</li>
<li>Update each parameter using the gradients, according to the optimization algorithm</li>
</ol>
</li>
</ol>
<p>Then, given a new data point, you can use the model to predict its class.</p>
<p>The initialization step can be critical to the model’s ultimate performance, and it requires the right method. To illustrate this, consider the three-layer neural network below. You can try initializing this network with different methods and observe the impact on the learning.</p>
<div>
<div>
<h3>1. Choose input dataset</h3>
<p>Select a training dataset.</p>
<p></p>
<p>This legend details the color scheme for labels, and the values of the weights/gradients.</p>
<p></p>
</div>
<div>
<h3>3. Train the network.</h3>
<p>Observe the cost function and the decision boundary.</p>
<p></p>
<p></p>
<p></p>
<p></p>
</div>
</div>
<p>What do you notice about the gradients and weights when the initialization method is zero?</p>
<blockquote>
<p>Initializing all the weights with zeros leads the neurons to learn the same features during training.</p>
</blockquote>
<p>In fact, any constant initialization scheme will perform very poorly. Consider a <span>neural network</span> with two hidden units, and assume we initialize all the biases to 0 and the weights with some constant $\alpha$. If we forward propagate an input $(x_1,x_2)$ in this network, the output of both hidden units will be $relu(\alpha x_1 + \alpha x_2)$. Thus, both hidden units will have identical influence on the cost, which will lead to identical gradients. Thus, both neurons will evolve symmetrically throughout training, effectively preventing different neurons from learning different things.</p>
<p>What do you notice about the cost plot when you initialize weights with values too small or too large?</p>
<blockquote>
<p>Despite breaking the symmetry, initializing the weights with values (i) too small or (ii) too large leads respectively to (i) slow learning or (ii) divergence.</p>
</blockquote>
<p>Choosing proper values for initialization is necessary for efficient training. We will investigate this further in the next section.</p>
<h1>II   The problem of exploding or vanishing gradients</h1>
<p>Consider this 9-layer neural network.</p>
<p><img alt="9 layer" title="9 layer" src="https://www.deeplearning.ai/ai-notes/assets/images/article/initialization/9layer.png"></p>
<p>At every iteration of the optimization loop (forward, cost, backward, update), we observe that backpropagated gradients are either amplified or minimized as you move from the output layer towards the input layer. This result makes sense if you consider the following example.</p>
<p>Assume all the activation functions are linear (identity function). Then the output activation is:</p>
<p>$$
\hat{y} = a^{[L]} = W^{[L]}W^{[L-1]}W^{[L-2]}\dots W^{[3]}W^{[2]}W^{[1]}x
$$</p>
<p>where $L=10$ and $W^{[1]},W^{[2]},\dots,W^{[L-1]}$ are all matrices of size $(2,2)$ because layers $[1]$ to $[L-1]$ have 2 neurons and receive 2 inputs. With this in mind, and for illustrative purposes, if we assume $W^{[1]} = W^{[2]} = \dots = W^{[L-1]} = W$ the output prediction is $\hat{y} = W^{[L]}W^{L-1}x$ (where $W^{L-1}$ takes the matrix $W$ to the power of $L-1$, while $W^{[L]}$ denotes the $L^{th}$ matrix).</p>
<p>What would be the outcome of initialization values that were too small, too large or appropriate?</p>
<h3>Case 1: A too-large initialization leads to exploding gradients</h3>
<p>Consider the case where every weight is initialized slightly larger than the identity matrix.</p>
<p>$$
W^{[1]} = W^{[2]} = \dots = W^{[L-1]}=\begin{bmatrix}1.5 &amp; 0 \\ 0 &amp; 1.5\end{bmatrix}
$$</p>
<p>This simplifies to $\hat{y} = W^{[L]}1.5^{L-1}x$, and the values of $a^{[l]}$ increase exponentially with $l$. When these activations are used in backward propagation, this leads to the exploding gradient problem. That is, the gradients of the cost with the respect to the parameters are too big. This leads the cost to oscillate around its minimum value.</p>
<h3>Case 2: A too-small initialization leads to vanishing gradients</h3>
<p>Similarly, consider the case where every weight is initialized slightly smaller than the identity matrix.</p>
<p>$$
W^{[1]} = W^{[2]} = \dots = W^{[L-1]}=\begin{bmatrix}0.5 &amp; 0 \\ 0 &amp; 0.5\end{bmatrix}
$$</p>
<p>This simplifies to $\hat{y} = W^{[L]}0.5^{L-1}x$, and the values of the activation $a^{[l]}$ decrease exponentially with $l$. When these activations are used in backward propagation, this leads to the vanishing gradient problem. The gradients of the cost with respect to the parameters are too small, leading to convergence of the cost before it has reached the minimum value.</p>
<p>All in all, initializing weights with inappropriate values will lead to divergence or a slow-down in the training of your neural network. Although we illustrated the exploding/vanishing gradient problem with simple symmetrical weight matrices, the observation generalizes to any initialization values that are too small or too large.</p>
<h1>III   How to find appropriate initialization values</h1>
<p>To prevent the gradients of the network’s activations from vanishing or exploding, we will stick to the following rules of thumb:</p>
<ol>
<li>The <span>mean</span> of the activations should be zero.</li>
<li>The <span>variance</span> of the activations should stay the same across every layer.</li>
</ol>
<p>Under these two assumptions, the backpropagated gradient signal should not be multiplied by values too small or too large in any layer. It should travel to the input layer without exploding or vanishing.</p>
<p>More concretely, consider a <span>layer $l$</span>. Its forward propagation is:</p>
<p>$$
\begin{aligned}a^{[l-1]} &amp;= g^{[l-1]}(z^{[l-1]})\\ z^{[l]} &amp;= W^{[l]}a^{[l-1]} + b^{[l]}\\ a^{[l]} &amp;= g^{[l]}(z^{[l]})\end{aligned}
$$</p>
<p>We would like the following to hold:</p>
<p>$$
\begin{aligned}E[a^{[l-1]}] &amp;= E[a^{[l]}]\\ Var(a^{[l-1]}) &amp;= Var(a^{[l]})\end{aligned}
$$</p>
<p>Ensuring zero-mean and maintaining the value of the variance of the input of every layer guarantees no exploding/vanishing signal, as we’ll explain in a moment. This method applies both to the forward propagation (for activations) and backward propagation (for gradients of the cost with respect to activations). The recommended initialization is Xavier initialization (or one of its derived methods), for every layer $l$:</p>
<p>$$
\begin{aligned}W^{[l]} &amp;\sim \mathcal{N}(\mu=0,\sigma^2 = \frac{1}{n^{[l-1]}})\\ b^{[l]} &amp;= 0\end{aligned}
$$</p>
<p>In other words, all the weights of layer $l$ are picked randomly from a <span>normal distribution</span> with mean $\mu = 0$ and variance $\sigma^2 = \frac{1}{n^{[l-1]}}$ where $n^{[l-1]}$ is the number of neuron in layer $l-1$. Biases are initialized with zeros.</p>
<p>The visualization below illustrates the influence of the Xavier initialization on each layer’s activations for a five-layer fully-connected neural network.</p>
<div>
<div>
<h3>1. Load your dataset</h3>
<p>Load 10,000 handwritten digits images (<a href="http://yann.lecun.com/exdb/mnist/">MNIST</a>).</p> </div>
<div>
<p>Input batch of 100 images</p>
<p></p>
<label>
Batch: <span>0</span> </label>
<label>
Epoch: <span>0</span>
</label>
</div>
<div>
<p>Output predictions of 100 images</p>
<p></p>
<label>
Misclassified: <span>0/100</span>
</label>
<label>
Cost: <span>0.00</span>
</label> </div> </div>
<p>You can find the theory behind this visualization in <a href="http://proceedings.mlr.press/v9/glorot10a/glorot10a.pdf?hc_location=ufi">Glorot et al. (2010)</a>. The next section presents the mathematical justification for Xavier initialization and explains more precisely why it is an effective initialization.</p>
<h1>IV   Justification for Xavier initialization</h1>
<p>In this section, we will show that Xavier Initialization keeps the variance the same across every layer. We will assume that our layer’s activations are normally distributed around zero. Sometimes it helps to understand the mathematical justification to grasp the concept, but you can understand the fundamental idea without the math.</p>
<p>Let’s work on the <span>layer $l$</span> described in part (III) and assume the activation function is <span>$tanh$</span>. The forward propagation is:</p>
<p>$$
\begin{aligned} z^{[l]} &amp;= W^{[l]}a^{[l-1]} + b^{[l]} \\ a^{[l]} &amp;= tanh(z^{[l]}) \end{aligned}
$$</p>
<p>The goal is to derive a relationship between $Var(a^{[l-1]})$ and $Var(a^{[l]})$. We will then understand how we should initialize our weights such that: $Var(a^{[l-1]}) = Var(a^{[l]})$.</p>
<p>Assume we initialized our network with appropriate values and the input is normalized. Early on in the training, we are in the <span>linear regime</span> of $tanh$. Values are small enough and thus $tanh(z^{[l]})\approx z^{[l]}$, meaning that:</p>
<p>$$
Var(a^{[l]}) = Var(z^{[l]})
$$</p>
<p>Moreover, $z^{[l]} = W^{[l]}a^{[l-1]} + b^{[l]} = vector(z_1^{[l]},z_2^{[l]},\dots,z_{n^{[l]}}^{[l]})$ where $z_k^{[l]} = \sum_{j=1}^{n^{[l-1]}}w_{kj}^{[l]}a_j^{[l-1]} + b_k^{[l]}$. For simplicity, let’s assume that $b^{[l]} = 0$ (it will end up being true given the choice of initialization we will choose). Thus, looking <span>element-wise</span> at the previous equation $Var(a^{[l-1]}) = Var(a^{[l]})$ now gives:</p>
<p>$$
Var(a_k^{[l]}) = Var(z_k^{[l]}) = Var(\sum_{j=1}^{n^{[l-1]}}w_{kj}^{[l]}a_j^{[l-1]})
$$</p>
<p>A common math trick is to extract the summation outside the variance. To do this, we must make the following three <span>assumptions</span>:</p>
<ol>
<li>Weights are independent and identically distributed</li>
<li>Inputs are independent and identically distributed</li>
<li>Weights and inputs are mutually independent</li>
</ol>
<p>Thus, now we have:</p>
<p>$$
Var(a_k^{[l]}) = Var(z_k^{[l]}) = Var(\sum_{j=1}^{n^{[l-1]}}w_{kj}^{[l]}a_j^{[l-1]}) = \sum_{j=1}^{n^{[l-1]}}Var(w_{kj}^{[l]}a_j^{[l-1]})
$$</p>
<p>Another common math trick is to convert the variance of a product into a product of variances. Here is the <span>formula</span> for it:</p>
<p>$$
Var(XY) = E[X]^2Var(Y) + Var(X)E[Y]^2 + Var(X)Var(Y)
$$</p>
<p>Using this formula with $X = w_{kj}^{[l]}$ and $Y = a_j^{[l-1]}$, we get:</p>
<p>$$
Var(w_{kj}^{[l]}a_j^{[l-1]}) = E[w_{kj}^{[l]}]^2Var(a_j^{[l-1]}) + Var(w_{kj}^{[l]})E[a_j^{[l-1]}]^2 + Var(w_{kj}^{[l]})Var(a_j^{[l-1]})
$$</p>
<p>We’re almost done! The first assumption leads to $E[w_{kj}^{[l]}]^2 = 0$ and the second assumption leads to $E[a_j^{[l-1]}]^2 = 0$ because weights are initialized with zero mean, and inputs are normalized. Thus:</p>
<p>$$
Var(z_k^{[l]}) = \sum_{j=1}^{n^{[l-1]}}Var(w_{kj}^{[l]})Var(a_j^{[l-1]}) = \sum_{j=1}^{n^{[l-1]}}Var(W^{[l]})Var(a^{[l-1]}) = n^{[l-1]}Var(W^{[l]})Var(a^{[l-1]})
$$</p>
<p>The equality above results from our first assumption stating that:</p>
<p>$$
Var(w_{kj}^{[l]}) = Var(w_{11}^{[l]}) = Var(w_{12}^{[l]})=\dots = Var(W^{[l]})
$$</p>
<p>Similarly the second assumption leads to:</p>
<p>$$
Var(a_j^{[l-1]}) = Var(a_1^{[l-1]}) = Var(a_2^{[l-1]})=\dots = Var(a^{[l-1]})
$$</p>
<p>With the same idea:</p>
<p>$$
Var(z^{[l]}) = Var(z_k^{[l]})
$$</p>
<p>Wrapping up everything, we have:</p>
<p>$$
Var(a^{[l]}) = n^{[l-1]}Var(W^{[l]})Var(a^{[l-1]})
$$</p>
<p>Voilà! If we want the variance to stay the same across layers ($Var(a^{[l]}) = Var(a^{[l-1]})$), we need $Var(W^{[l]}) = \frac{1}{n^{[l-1]}}$. This justifies the choice of variance for Xavier initialization.</p>
<p>Notice that in the previous steps we did not choose a specific layer $l$. Thus, we have shown that this expression holds for every layer of our network. Let $L$ be the output layer of our network. Using this expression at every layer, we can link the output layer’s variance to the input layer’s variance:</p>
<p>$$
\begin{aligned} Var(a^{[L]}) &amp;= n^{[L-1]}Var(W^{[L]})Var(a^{[L-1]}) \\ &amp;= n^{[L-1]}Var(W^{[L]})n^{[L-2]}Var(W^{[L-1]})Var(a^{[L-2]})\\ &amp;=\dots\\ &amp;= \left[\prod_{l=1}^L n^{[l-1]}Var(W^{[l]})\right]Var(x)\end{aligned}
$$</p>
<p>Depending on how we initialize our weights, the relationship between the variance of our output and input will vary dramatically. Notice the following three cases.</p>
<p>$$
n^{[l-1]}Var(W^{[l]}) \begin{cases} &lt; 1 &amp;\implies \text{Vanishing Signal}\\ = 1 &amp; \implies Var(a^{[L]}) = Var(x)\\ &gt; 1 &amp; \implies \text{Exploding Signal}\end{cases}
$$</p>
<p>Thus, in order to avoid the vanishing or exploding of the forward propagated signal, we must set $n^{[l-1]}Var(W^{[l]}) = 1$ by initializing $Var(W^{[l]}) = \frac{1}{n^{[l-1]}}$.</p>
<p>Throughout the justification, we worked on activations computed during the forward propagation. The same result can be derived for the backpropagated gradients. Doing so, you will see that in order to avoid the vanishing or exploding gradient problem, we must set $n^{[l]}Var(W^{[l]}) = 1$ by initializing $Var(W^{[l]}) = \frac{1}{n^{[l]}}$.</p>
<h1>Conclusion</h1>
<p>In practice, Machine Learning Engineers using Xavier initialization would either initialize the weights as $\mathcal{N}(0,\frac{1}{n^{[l-1]}})$ or as $\mathcal{N}(0,\frac{2}{n^{[l-1]} + n^{[l]}})$. The variance term of the latter distribution is the harmonic mean of $\frac{1}{n^{[l-1]}}$ and $\frac{1}{n^{[l]}}$.</p>
<p>This is a theoretical justification for Xavier initialization. Xavier initialization works with tanh activations. Myriad other initialization methods exist. If you are using ReLU, for example, a common initialization is He initialization (<a href="https://arxiv.org/pdf/1502.01852.pdf">He et al., Delving Deep into Rectifiers</a>), in which the weights are initialized by multiplying by 2 the variance of the Xavier initialization. While the justification for this initialization is slightly more complicated, it follows the same thought process as the one for tanh.</p>
<div>
<p><strong>Learn more about how to effectively initialize parameters in</strong><br><strong>Course 2 of the Deep Learning Specialization</strong></p>
<p><a href="http://bit.ly/2VMCWcR" target="_blank">Enroll now</a></p>
</div>
<div>
<p>Examples include Adam, Momentum, RMSProp, Stochastic and Batch Gradient Descent methods.</p>
</div>
<div>
<p>A neural network with two hidden relu units and a sigmoid output unit.</p>
</div>
<div>
<p>Mean is a measure of the center or expectation of a random variable.</p>
</div>
<div>
<p>Variance is a measure of how much a random variable is spread around its mean. In deep learning, the random variable could be the data, the prediction, the weights, the activations, etc.</p>
</div>
<div>
<p><img src="https://www.deeplearning.ai/ai-notes/assets/images/article/initialization/layerl.png"> $a^{[l-1]}$ represents the input to layer $l$ and $a^{[l]}$ represents the output. $g^{[l]}$ is the activation function of layer $l$. $n^{[l]}$ is the number of neuron in layer $l$.</p>
</div>
<div>
<p><img src="https://www.deeplearning.ai/ai-notes/assets/images/article/initialization/normal.png"> Values generated from a normal distribution $\mathcal{N}(\mu,\sigma^2)$ are symmetric around the mean $\mu$.</p>
</div>
<div>
<p><img src="https://www.deeplearning.ai/ai-notes/assets/images/article/initialization/layerl.png"> $a^{[l-1]}$ represents the input to layer $l$ and $a^{[l]}$ represents the output.</p>
</div>
<div>
<p>$tanh$ is a non-linear function defined as $tanh(x) = \frac{1 - e^{-2x}}{1 + e^{-2x}}$.</p>
</div>
<div>
<p>Important properties of $tanh$ are its parity ($tanh(-x) = -tanh(x)$) and its linearity around 0 ($tanh'(0) = 1$).</p>
</div>
<div>
<p>The variance of the vector is the same as the variance of any of its entries, because all its entries are drawn independently and identically from the same distribution (i.i.d.).</p>
</div>
<div>
<p>These assumptions are not always true, but they are necessary to approach the problem theoretically at this point.</p>
</div>
<div>
<p>This is only true for independent random variables.</p>
</div>
</section></div></div></div></div><br>
<br>
<br>
<br>
<br>
<br>
via deeplearning.ai <br>
<br>
January 30, 2024 at 01:54PM
